client/v3/config_test.go

```diff
@@ -0,0 +1,159 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clientv3
+
+import (
+	"crypto/tls"
+	"testing"
+	"time"
+
+	"go.uber.org/zap"
+)
+
+func TestNewClientConfig(t *testing.T) {
+	cases := []struct {
+		name string
+		spec ConfigSpec
+		conf Config
+	}{
+		{
+			name: "only has basic info",
+			spec: ConfigSpec{
+				Endpoints:        []string{"http://192.168.0.10:2379"},
+				DialTimeout:      2 * time.Second,
+				KeepAliveTime:    3 * time.Second,
+				KeepAliveTimeout: 5 * time.Second,
+			},
+			conf: Config{
+				Endpoints:            []string{"http://192.168.0.10:2379"},
+				DialTimeout:          2 * time.Second,
+				DialKeepAliveTime:    3 * time.Second,
+				DialKeepAliveTimeout: 5 * time.Second,
+			},
+		},
+		{
+			name: "auth enabled",
+			spec: ConfigSpec{
+				Endpoints:        []string{"http://192.168.0.12:2379"},
+				DialTimeout:      1 * time.Second,
+				KeepAliveTime:    4 * time.Second,
+				KeepAliveTimeout: 6 * time.Second,
+				Auth: &AuthConfig{
+					Username: "test",
+					Password: "changeme",
+				},
+			},
+			conf: Config{
+				Endpoints:            []string{"http://192.168.0.12:2379"},
+				DialTimeout:          1 * time.Second,
+				DialKeepAliveTime:    4 * time.Second,
+				DialKeepAliveTimeout: 6 * time.Second,
+				Username:             "test",
+				Password:             "changeme",
+			},
+		},
+		{
+			name: "default secure transport",
+			spec: ConfigSpec{
+				Endpoints:        []string{"http://192.168.0.10:2379"},
+				DialTimeout:      2 * time.Second,
+				KeepAliveTime:    3 * time.Second,
+				KeepAliveTimeout: 5 * time.Second,
+				Secure: &SecureConfig{
+					InsecureTransport: false,
+				},
+			},
+			conf: Config{
+				Endpoints:            []string{"http://192.168.0.10:2379"},
+				DialTimeout:          2 * time.Second,
+				DialKeepAliveTime:    3 * time.Second,
+				DialKeepAliveTimeout: 5 * time.Second,
+				TLS:                  &tls.Config{},
+			},
+		},
+		{
+			name: "default secure transport and skip TLS verification",
+			spec: ConfigSpec{
+				Endpoints:        []string{"http://192.168.0.13:2379"},
+				DialTimeout:      1 * time.Second,
+				KeepAliveTime:    3 * time.Second,
+				KeepAliveTimeout: 5 * time.Second,
+				Secure: &SecureConfig{
+					InsecureTransport:  false,
+					InsecureSkipVerify: true,
+				},
+			},
+			conf: Config{
+				Endpoints:            []string{"http://192.168.0.13:2379"},
+				DialTimeout:          1 * time.Second,
+				DialKeepAliveTime:    3 * time.Second,
+				DialKeepAliveTimeout: 5 * time.Second,
+				TLS: &tls.Config{
+					InsecureSkipVerify: true,
+				},
+			},
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			lg, _ := zap.NewProduction()
+
+			cfg, err := NewClientConfig(&tc.spec, lg)
+			if err != nil {
+				t.Fatalf("Unexpected error: %v", err)
+			}
+
+			// check endpoints
```

Comment:
Have you considered to use testify `assert.Equal` instead of doing all comparisons yourself?

Reply:
It's a good point. Done & Thanks.